README.md
<img src="https://github.com/francescopisu/modelsight/blob/main/docs/images/modelsight-logo.svg" width="175" align="right" />

# Welcome to modelsight

> Better insights into Machine Learning models performance.

Modelsight is a collection of functions that create publication-ready graphics for the visual evaluation of cross-validated binary classifiers adhering to the scikit-learn interface. 

While the majority of `modelsight` functionalities are compatible with any binary classifier that follows scikit-learn's interface, it's worth noting that, as of now, some features are exclusive to ExplainableBoostingClassifier from the interpretml package.

| Overview | |
|---|---|
| **Open Source** | [![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0) |
| **CI/CD** | [![github-actions](https://img.shields.io/github/actions/workflow/status/francescopisu/modelsight/ci-cd.yml?logo=github)](https://github.com/francescopisu/modelsight/actions/workflows/ci-cd.yml) [![!codecov](https://img.shields.io/codecov/c/github/francescopisu/modelsight?label=codecov&logo=codecov)](https://codecov.io/gh/francescopisu/modelsight) [![readthedocs](https://img.shields.io/readthedocs/modelsight?logo=readthedocs)](https://modelsight.readthedocs.io/en/latest/) |
| **Code** |  [![!pypi](https://img.shields.io/pypi/v/modelsight?color=orange)](https://pypi.org/project/modelsight/) [![!python-versions](https://img.shields.io/pypi/pyversions/modelsight)](https://www.python.org/) |

## 💫 Features
Our objective is to enhance the visual assessment of cross-validated binary classifiers by introducing a set of functions tailored for generating publication-ready plots.

To effectively utilize modelsight, you should provide a dictionary where the keys represent model names, and the values consist of CVModellingOutput objects. CVModellingOutput is a custom data class designed to encapsulate the results of a cross-validation process for a specific binary classifier. For detailed information about CVModellingOutput, please refer to the [API reference](https://modelsight.readthedocs.io/en/latest/autoapi/modelsight/_typing/index.html#modelsight._typing.CVModellingOutput).

Before using the functionalities offered by this library, ensure that your cross-validation data adheres to this structure.

| Module | Status | Links |
|---|---|---|
| **[Calibration]** | maturing | [Tutorial](https://modelsight.readthedocs.io/en/latest/01_calibration.html) · [API Reference](https://modelsight.readthedocs.io/en/latest/autoapi/modelsight/calibration/index.html) |
| **[Curves]** | maturing | [Tutorial](https://modelsight.readthedocs.io/en/latest/02_curves.html) · [API Reference](https://modelsight.readthedocs.io/en/latest/autoapi/modelsight/curves/index.html) |

[calibration]: https://github.com/francescopisu/modelsight/tree/main/src/modelsight/calibration
[curves]: https://github.com/francescopisu/modelsight/tree/main/src/modelsight/curves

## :eyeglasses: Install modelsight
- **Operating system**: macOS X · Linux
- **Python version**: 3.10 (only 64-bit)
- **Package managers**: [pip]

[pip]: https://pip.pypa.io/en/stable/

### pip
Using pip, modelsight releases are available as source packages and binary wheels. You can see all available wheels [here](https://pypi.org/simple/modelsight/).
```console
$ pip install modelsight
```

## :zap: Quickstart
The following code will cross-validate five different binary classifiers and produce a dictionary of CVModellingOutput.

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold, RepeatedStratifiedKFold
from sklearn.metrics import roc_curve
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from interpret.glassbox import ExplainableBoostingClassifier

from tests.utils import (
    select_features, 
    get_feature_selector, 
    get_calibrated_model
)

from modelsight._typing import CVModellingOutput, Estimator

# Load data
X, y = load_breast_cancer(return_X_y=True, as_frame=True)

# Define factory for binary classifiers
class BinaryModelFactory:
    def get_model(self, model_name: str) -> Estimator:
        if model_name == "EBM":
            return ExplainableBoostingClassifier(random_state=cv_config.get("SEED"),
                                        interactions=6,
                                        learning_rate=0.02,
                                        min_samples_leaf=5,
                                        n_jobs=-1)
        elif model_name == "LR":
            return LogisticRegression(penalty="elasticnet",
                                 solver="saga",
                                 l1_ratio=0.3,
                                 max_iter=10000,
                                 random_state=cv_config.get("SEED"))
        elif model_name == "SVC": 
            return SVC(probability=True, 
                   class_weight="balanced", 
                   random_state=cv_config.get("SEED"))
        elif model_name == "RF": 
            return RandomForestClassifier(random_state=cv_config.get("SEED"),
                                     n_estimators=5,
                                     max_depth=3,
                                     n_jobs=-1)
        elif model_name == "KNN": 
            return KNeighborsClassifier(n_jobs=-1)        
        else:
            raise ValueError(f"{model_name} is not a valid estimator name.")

cv_config = {
    "N_REPEATS": 2,
    "N_SPLITS": 10,
    "SHUFFLE": True,
    "SCALE": False,
    "CALIBRATE": True,
    "CALIB_FRACTION": 0.15,
    "SEED": 1303
}

# define outer and inner cross-validation schemes
outer_cv = RepeatedStratifiedKFold(n_repeats=cv_config.get("N_REPEATS"),
                                   n_splits=cv_config.get("N_SPLITS"),
                                   random_state=cv_config.get("SEED"))

inner_cv = StratifiedKFold(n_splits=cv_config.get("N_SPLITS"),
                           shuffle=cv_config.get("SHUFFLE"),
                           random_state=cv_config.get("SEED"))

# define models names, factory and cv results dictionary
models_names = ["EBM", "LR", "SVC", "RF", "KNN"]
model_factory = BinaryModelFactory()
cv_results = dict()

for model_name in models_names:
    print(f"Processing model {model_name}\n")

    gts_train = []
    gts_val = []
    probas_train = []
    probas_val = []
    gts_train_conc = []
    gts_val_conc = []
    probas_train_conc = []
    probas_val_conc = []

    models = []
    errors = []
    correct = []
    features = []
        
    for i, (train_idx, val_idx) in enumerate(outer_cv.split(X, y)):
        Xtemp, ytemp = X.iloc[train_idx, :], y.iloc[train_idx]
        Xval, yval = X.iloc[val_idx, :], y.iloc[val_idx]

        if cv_config.get("CALIBRATE"):
            Xtrain, Xcal, ytrain, ycal = train_test_split(Xtemp, ytemp,
                                                          test_size=cv_config.get(
                                                              "CALIB_FRACTION"),
                                                          stratify=ytemp,
                                                          random_state=cv_config.get("SEED"))
        else:
            Xtrain, ytrain = Xtemp, ytemp

        model = model_factory.get_model(model_name)

        # select features
        feat_subset = select_features(Xtrain, ytrain,
                                      selector=get_feature_selector(
                                          cv_config.get("SEED")),
                                      cv=inner_cv,
                                      scale=False,
                                      frac=1.25)
        features.append(feat_subset)

        if cv_config.get("SCALE"):
            numeric_cols = Xtrain.select_dtypes(
                include=[np.float64, np.int64]).columns.tolist()
            scaler = StandardScaler()
            Xtrain.loc[:, numeric_cols] = scaler.fit_transform(
                Xtrain.loc[:, numeric_cols])
            Xtest.loc[:, numeric_cols] = scaler.transform(
                Xtest.loc[:, numeric_cols])

        model.fit(Xtrain.loc[:, feat_subset], ytrain)

        if cv_config.get("CALIBRATE"):
            model = get_calibrated_model(model,
                                         X=Xcal.loc[:, feat_subset],
                                         y=ycal)

        models.append(model)

        # accumulate ground-truths
        gts_train.append(ytrain)
        gts_val.append(yval)

        # accumulate predictions
        train_pred_probas = model.predict_proba(Xtrain.loc[:, feat_subset])[:, 1]
        val_pred_probas = model.predict_proba(Xval.loc[:, feat_subset])[:, 1]

        probas_train.append(train_pred_probas)
        probas_val.append(val_pred_probas)

        # identify correct and erroneous predictions according to the
        # classification cut-off that maximizes the Youden's J index
        fpr, tpr, thresholds = roc_curve(ytrain, train_pred_probas)
        idx = np.argmax(tpr - fpr)
        youden = thresholds[idx]

        labels_val = np.where(val_pred_probas >= youden, 1, 0)

        # indexes of validation instances misclassified by the model
        error_idxs = Xval[(yval != labels_val)].index
        errors.append(error_idxs)

        # indexes of correct predictions
        correct.append(Xval[(yval == labels_val)].index)

    # CV results for current model
    curr_est_results = CVModellingOutput(
        gts_train=gts_train,
        gts_val=gts_val,
        probas_train=probas_train,
        probas_val=probas_val,
        gts_train_conc=np.concatenate(gts_train),
        gts_val_conc=np.concatenate(gts_val),
        probas_train_conc=np.concatenate(probas_train),
        probas_val_conc=np.concatenate(probas_val),
        models=models,
        errors=errors,
        correct=correct,
        features=features
    )
    
    cv_results[model_name] = curr_est_results
```

# Plot the average receiver-operating characteristic (ROC) curves
```python
from modelsight.curves import (
    average_roc_curves, 
    roc_comparisons, 
    add_annotations
)

model_names = list(cv_results.keys())
alpha = 0.05
alph_str = str(alpha).split(".")[1]
alpha_formatted = f".{alph_str}"
roc_symbol = "*"
palette = [Colors.green2, Colors.blue, Colors.yellow, Colors.violet, Colors.darksalmon]
n_boot = 100

f, ax = plt.subplots(1, 1, figsize=(8, 8))

kwargs = dict()

f, ax, barplot, bars, all_data = average_roc_curves(cv_results,
                                                    colors=palette,
                                                    model_keys=model_names,
                                                    show_ci=True,
                                                    n_boot=n_boot,
                                                    bars_pos=[
                                                        0.3, 0.01, 0.6, 0.075*len(model_names)],
                                                    random_state=cv_config.get("SEED"),
                                                    ax=ax,
                                                    **kwargs)

roc_comparisons_results = roc_comparisons(cv_results, "EBM")

kwargs = dict(space_between_whiskers = 0.07)
order = [
    ("EBM", "RF"),
    ("EBM", "SVC"),
    ("EBM", "LR"),
    ("EBM", "KNN")
]
ax_annot = add_annotations(roc_comparisons_results, 
                  alpha = 0.05, 
                  bars=bars, 
                  direction = "vertical",
                  order = order,
                  symbol = roc_symbol,
                  symbol_fontsize = 30,
                  voffset = -0.05,
                  ext_voffset=0,
                  ext_hoffset=0,
                  ax=barplot,
                  **kwargs)
```
This produces the following plot:
<p align="center">
<img src='https://github.com/francescopisu/modelsight/assets/14362976/a948c6e6-ee49-4754-aa64-4a43ffe4050c' alt="Average ROC plot" width="600"/>
</p>

### Assess calibration of the ExplaianbleBoostingClassifier across 
We will now compute median Brier score (95% CI) of the ExplainableBoostingClassifier and use it to annotate the calibration plot.
```python
from sklearn.metrics import brier_score_loss
from modelsight.calibration import hosmer_lemeshow_plot

briers = []
for gt, preds in zip(cv_results["EBM"].gts_val, cv_results["EBM"].probas_val):
    brier = brier_score_loss(gt, preds)
    briers.append(brier)

brier_low, brier_med, brier_up = np.percentile(briers, [2.5, 50, 97.5])

brier_annot = f"{brier_med:.2f} ({brier_low:.2f} - {brier_up:.2f})"

f, ax = plt.subplots(1, 1, figsize=(11,6))

f, ax = hosmer_lemeshow_plot(cv_results["EBM"].gts_val_conc,
                             cv_results["EBM"].probas_val_conc,
                             n_bins=10,
                             colors=(Colors.darksalmon, Colors.green2),
                             annotate_bars=True,
                             title="",
                             brier_score_annot=brier_annot,
                             ax=ax
                             )
```
This produces the following plot:
<p align="center">
<img src='https://github.com/francescopisu/modelsight/assets/14362976/af435cb6-cd88-4b00-996a-468f59c525c0' alt="Average ROC plot" width="600"/>
</p>


## :gift_heart: Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## 🛣️ Roadmap
Features:
- [x] Calibration module to assess calibration of ML predicted probabilities via Hosmer-Lemeshow plot
- [x] Average Receiver-operating characteristic curves
- [ ] Average Precision-recall curves
- [ ] Feature importance (Global explanation)
- [ ] Individualized explanations (Local explanation)


CHANGELOG.md
# CHANGELOG



## v0.3.7 (2023-09-24)

### Build

* build: update python-semantic-release ([`214ff7e`](https://github.com/francescopisu/modelsight/commit/214ff7edbc0380522fcacb0dd83f4120ce8c7970))

### Fix

* fix: replace references to src folder with modelsight folder ([`70471d2`](https://github.com/francescopisu/modelsight/commit/70471d26c34bacb550123658ff8236920671ba64))

* fix: remove src folder ([`1cf5d37`](https://github.com/francescopisu/modelsight/commit/1cf5d3768b184238769acfa5bffa5ef80c688a7c))

### Unknown

* Merge pull request #18 from francescopisu/fix/remove-src-folder

fix: remove src folder ([`6441f81`](https://github.com/francescopisu/modelsight/commit/6441f8164a8fc4ff8d5ea784c8be2eea71855b70))


## v0.3.6 (2023-09-24)

### Documentation

* docs: improve docstring docs #13 ([`fec818b`](https://github.com/francescopisu/modelsight/commit/fec818b68a88ff8f71b29e3438572ad300e6161c))

* docs: improve docstring docs #13 ([`9fc34fd`](https://github.com/francescopisu/modelsight/commit/9fc34fd5276d33c920653b847d112dc03c9ba92d))

### Unknown

* Merge pull request #16 from francescopisu/dev

[sync] Merge dev to main ([`57d8381`](https://github.com/francescopisu/modelsight/commit/57d83818693fbc1df773f0d53e5086f69d68a0b3))


## v0.3.5 (2023-09-17)

### Documentation

* docs: update README with graphical examples #10 ([`26102c4`](https://github.com/francescopisu/modelsight/commit/26102c4a4091238675e7270e3c16936c28a84682))

* docs: update README with graphical examples #10 ([`0cabbb0`](https://github.com/francescopisu/modelsight/commit/0cabbb02a2b48617110bac0738a2c2b9eeb144dc))


## v0.3.4 (2023-09-17)

### Documentation

* docs: update example notebook + self-contained python script #3 ([`47d4cbc`](https://github.com/francescopisu/modelsight/commit/47d4cbc859378e7e030bdae2ffeb69d041964f63))

* docs: update example notebook with average ROC curves plot of multiple binary classifiers ([`ef14075`](https://github.com/francescopisu/modelsight/commit/ef14075f126b02147b9b4aaab07e7b38e959ab72))

### Unknown

* Merge pull request #9 from francescopisu/dev

Dev ([`a6374ec`](https://github.com/francescopisu/modelsight/commit/a6374ec511d92ffc698f5edd95fd7867f5ea8bc4))

* Merge pull request #8 from francescopisu/docs/update-example-notebook

Update example notebook and add self-contained quickstart python script ([`77d168d`](https://github.com/francescopisu/modelsight/commit/77d168d360134f8d83b1655ed36cc46db4c7a8df))


## v0.3.3 (2023-09-15)

### Fix

* fix: numpy float deprecated alias #4 ([`3e71866`](https://github.com/francescopisu/modelsight/commit/3e71866ed365ea831c21cbd4dec2513ea08b3a97))

### Unknown

* Merge pull request #6 from francescopisu/dev

Dev ([`272404f`](https://github.com/francescopisu/modelsight/commit/272404f94ee2167f036dbc62ecb53ffdb852480d))

* Merge pull request #5 from francescopisu/fix/numpy-float

fix: numpy float deprecated alias #4 ([`5bf8cc3`](https://github.com/francescopisu/modelsight/commit/5bf8cc3e8dbba3f8e10f5f4b06d071e046fac716))


## v0.3.2 (2023-09-15)

### Build

* build: add CI workflow ([`f83e647`](https://github.com/francescopisu/modelsight/commit/f83e6471b7bfa8af67a7e4cf434905c3b251da52))

* build: fix security vulnerabilities ([`925d517`](https://github.com/francescopisu/modelsight/commit/925d517e8c75f169aef98438c35785ec059d483a))

* build: preparing for release v0.3.0 ([`341a11a`](https://github.com/francescopisu/modelsight/commit/341a11a5b719f11e7e03f414efb8f0d1db32df74))


## v0.3.0 (2023-09-14)

### Build

* build: trying to fix rtd compilation error ([`c825a6f`](https://github.com/francescopisu/modelsight/commit/c825a6f66444fb3f88f35d55837a4d6836760703))

* build: trying to fix rtd compilation error ([`c1b95a8`](https://github.com/francescopisu/modelsight/commit/c1b95a884919b3256eb6e2c71efea01774e1cd2a))

* build: trying to fix rtd compilation error ([`aabba19`](https://github.com/francescopisu/modelsight/commit/aabba19da56be000501c0fc1791122fe8bbd2a1e))

* build: specified correct python version for rtd ([`97911c9`](https://github.com/francescopisu/modelsight/commit/97911c98038cbffc963e34d1c83728f520302157))

* build: preparing for release v0.2.0 ([`23b2aa5`](https://github.com/francescopisu/modelsight/commit/23b2aa5bec6d637bc1be0bc1ecf144c8ed2e8295))

* build: get current version by tag ([`72f860b`](https://github.com/francescopisu/modelsight/commit/72f860b47f1d0099fcc398c5014944bd089e20dd))

### Documentation

* docs: fix quickstart code ([`497770b`](https://github.com/francescopisu/modelsight/commit/497770be6873fa68f7599467cee8bb77972dce17))

* docs: fix logo path ([`a89feff`](https://github.com/francescopisu/modelsight/commit/a89feff408eb197fb4c5f10a2bddb1379ab4ffff))

* docs: update readme ([`b91d7ea`](https://github.com/francescopisu/modelsight/commit/b91d7ea1077d585a240862d398c3f42477e5f292))

### Feature

* feat: comparison between median AUCs + tests ([`2ca124c`](https://github.com/francescopisu/modelsight/commit/2ca124c305cfe5f972ce22c5202ea2b66d21dff0))

### Unknown

* Merge pull request #2 from francescopisu/dev

feat: comparison between median AUCs + tests ([`a46fd54`](https://github.com/francescopisu/modelsight/commit/a46fd543b4bb1515b56281c696300d08f221bcaa))

* Merge remote-tracking branch &#39;refs/remotes/origin/main&#39; ([`55b6827`](https://github.com/francescopisu/modelsight/commit/55b6827061fcb0f846f4f5bb4734c4d129ed47a9))


## v0.2.0 (2023-07-17)

### Documentation

* docs: update roadmap ([`1b9875d`](https://github.com/francescopisu/modelsight/commit/1b9875dc9e75c1e677c7f3a03189cf9fa1552257))

* docs: add roadmap to readme ([`f16718b`](https://github.com/francescopisu/modelsight/commit/f16718be186e8c896cd7fd7adc92d9d47b9e385f))

* docs: update readme and add notebook for example usage ([`6f13516`](https://github.com/francescopisu/modelsight/commit/6f13516d53a9b2beb8a47fd73ab771fbea7bca76))

### Feature


modelsight/curves/roc.py
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
from typing import List, Dict, Tuple
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve
from scikits.bootstrap import bootstrap_indices as scikits_boot_indices

from modelsight._typing import CVModellingOutput, SeedType


def average_roc_curves(cv_preds: Dict[str, CVModellingOutput],
                       colors: List[str],
                       model_keys_map: Dict[str, str] = {},
                       show_ci: bool = True,
                       n_boot: int = 1000,
                       bars_pos: Tuple[int, int, int, int] = (0.41, 0.01, 0.53, 0.30),
                       random_state: SeedType = 1234,
                       ax: plt.Axes = None,
                       **kwargs) -> Tuple[plt.Axes,
                                          plt.Axes,
                                          matplotlib.container.BarContainer,
                                          Dict[str, Dict[str, float]]]:
    """
    Generate receiver-operating characteristic curves for each model in cv_preds.

    Parameters
    ----------
    cv_preds: Dict[str, CVModellingOutput]
        A dictionary containing model-specific cross-validation modelling outputs.
    colors: List[str]
        A list of colors that will be used to color both curves and bars.
    model_keys_map: Dict[str, str] (default = {})
        A dictionary mapping model keys to model names.
    show_ci: bool (default = True)
        Whether bootstrapped confidence bands around curves should be shown.
    n_boot: int (default = 1000)
        Number of bootstrap iterations for generating confidence bands.
    bars_pos: Tuple[int, int, int, int]
        A tuple of four integers specifying the shape and position of the bar plot inset.
        (x position, y position, width, height)
    random_state: Seed (default = 1234)
        A seed for reproducibility.
    ax: plt.Axes (default = None)
        Optional Axes to plot curves onto.
    **kwargs:
        model_names_in_black: List[str]
            Names of models to show in black color, default is []

    Returns
    -------
    Tuple[plt.Axes, plt.Axes, matplotlib.container.BarContainer, Dict[str, Dict[str, float]]]
        First: the Axes containing the general plot.
        Second: the axes containing the bar plot inset.
        Third: the actual BarContainer of the bar plot inset.
        Fourth: A dictionary containing median (95%CI) area-under-curve over cross-validation
                for each model.
    """
    if not ax:
        fig, ax = plt.subplots(figsize=(8, 4))
    else:
        fig = ax.get_figure()
    
    rng = np.random.RandomState(random_state)
    auc_cis = dict()
    for j, (algo_name, outer_cv_data) in enumerate(cv_preds.items()):
        aucs = []

        gts = outer_cv_data.gts_val
        probas = outer_cv_data.probas_val

        for split_gts, split_probas in zip(gts, probas):
            auc_val = roc_auc_score(split_gts, split_probas)
            aucs.append(auc_val)

        fpr, tpr, thresholds = roc_curve(outer_cv_data.gts_val_conc,
                                         outer_cv_data.probas_val_conc)

        auc_low, auc_med, auc_up = np.percentile(aucs, [2.5, 50, 97.5])
        auc_cis[algo_name] = {"auc": auc_med,
                              "ci_low": auc_low, "ci_up": auc_up}

        ax.plot(fpr,
                tpr,
                linestyle='-',
                alpha=1.0,
                linewidth=2,
                color=colors[j])

        if show_ci:
            bootstrap_indices = list(scikits_boot_indices(data=outer_cv_data.probas_val_conc,
                                                          n_samples=n_boot,
                                                          seed=random_state))

            bootstrap_tprs = []
            for i in range(n_boot):
                sample_gt = outer_cv_data.gts_val_conc[bootstrap_indices[i]]
                sample_pred = outer_cv_data.probas_val_conc[bootstrap_indices[i]]

                bootstrap_fpr, bootstrap_tpr, bootstrap_thresholds = roc_curve(
                    sample_gt, sample_pred)

                # interpolate the bootstrapped fpr using the fpr based on accumulated
                # ground-truths and predicted probabilities
                interp_tpr = np.interp(fpr, bootstrap_fpr, bootstrap_tpr)
                interp_tpr[0] = 0.0

                bootstrap_tprs.append(interp_tpr)

            bootstrap_tprs = np.stack(bootstrap_tprs)

            tpr_lower = np.percentile(bootstrap_tprs, 2.5, axis=0)
            tpr_upper = np.percentile(bootstrap_tprs, 97.5, axis=0)

            ax.fill_between(fpr, tpr_lower, tpr_upper,
                            alpha=0.13,
                            color=colors[j])

    ax.set_xlabel('1 - Specificity',
                  fontdict={"weight": "normal", "size": 26},
                  labelpad=20)
    ax.set_ylabel('Sensitivity',
                  fontdict={"weight": "normal", "size": 26},
                  labelpad=20)

    ax.xaxis.set_tick_params(labelsize=23)
    ax.yaxis.set_tick_params(labelsize=23)

    ax.plot([0, 1], [0, 1], linestyle='--', lw=1,
            color='gray', label="Baseline", alpha=.8)

    ins = ax.inset_axes(bars_pos)
    ins.spines['top'].set_visible(False)
    ins.spines['right'].set_visible(False)
    ins.spines['bottom'].set_visible(False)

    ins.get_xaxis().set_ticks([])

    model_names = [model_keys_map.get(k, k) for k in cv_preds.keys()]
    model_aucs = [v["auc"] for _, v in auc_cis.items()]
    model_cis_low = np.array([v["auc"] - v["ci_low"]
                             for _, v in auc_cis.items()])
    model_cis_up = np.array([v["ci_up"] - v["auc"]
                            for _, v in auc_cis.items()])

    all_data = list(zip(model_names, model_aucs,
                    model_cis_low, model_cis_up, colors))
    all_data.sort(key=lambda x: x[1], reverse=False)

    model_names = [t[0] for t in all_data]
    model_aucs = [t[1] for t in all_data]
    model_cis_low = [t[2] for t in all_data]
    model_cis_up = [t[3] for t in all_data]
    model_colors = [t[4] for t in all_data]

    bars = ins.barh(range(len(model_names)), model_aucs,
                    xerr=[model_cis_low, model_cis_up],
                    align="center",
                    color=model_colors,
                    capsize=3,
                    error_kw=dict(linewidth=1)
                    )
    ins.invert_yaxis()
    ins.set_yticks(range(len(model_names)), labels=model_names, fontsize=16)

    model_names_in_black = kwargs.pop("model_names_in_black", [])
    for i, (name, auc, bar) in enumerate(zip(model_names, model_aucs, ins.patches)):
        ins.annotate(f"{auc:.2f}",
                     (bar.get_width(),
                      bar.get_y() + bar.get_height() / 2), ha='left', va='center',
                     size=20,
                     xytext=(0.04, i), color="#fff" if name not in model_names_in_black else "#000",
                     textcoords='data'
                     )

    ins.set_title("AUC (bars) and 95% CI (whiskers)",
                  fontsize=18,
                  fontweight="bold",
                  position=(0.4, 0.5))

    return fig, ax, ins, bars, all_data
